src/etl/impl/CacheLoader.h

```diff
@@ -0,0 +1,433 @@
+                    cacheLoadStyle_ = LoadStyle::ASYNC;
+                if (boost::iequals(*entry, "none") or boost::iequals(*entry, "no"))
+                    cacheLoadStyle_ = LoadStyle::NOT_AT_ALL;
+            }
+
+            numCacheDiffs_ = cache.valueOr<size_t>("num_diffs", numCacheDiffs_);
+            numCacheMarkers_ = cache.valueOr<size_t>("num_markers", numCacheMarkers_);
+            cachePageFetchSize_ = cache.valueOr<size_t>("page_fetch_size", cachePageFetchSize_);
+
+            if (auto peers = cache.maybeArray("peers"); peers)
+            {
+                for (auto const& peer : *peers)
+                {
+                    auto ip = peer.value<std::string>("ip");
+                    auto port = peer.value<uint32_t>("port");
+
+                    // todo: use emplace_back when clang is ready
+                    clioPeers_.push_back({ip, port});
+                }
+
+                unsigned seed = std::chrono::system_clock::now().time_since_epoch().count();
+                std::shuffle(std::begin(clioPeers_), std::end(clioPeers_), std::default_random_engine(seed));
+            }
+        }
+    }
+
+    ~CacheLoader()
+    {
+        stop();
+        if (thread_.joinable())
+            thread_.join();
+    }
+
+    /**
+     * @brief Populates the cache by walking through the given ledger.
+     *
+     * Should only be called once. The default behavior is to return immediately and populate the cache in the
+     * background. This can be overridden via config parameter, to populate synchronously, or not at all.
+     */
+    void
+    load(uint32_t seq)
+    {
+        if (cacheLoadStyle_ == LoadStyle::NOT_AT_ALL)
+        {
+            cache_.get().setDisabled();
+            log_.warn() << "Cache is disabled. Not loading";
+            return;
+        }
+
+        if (cache_.get().isFull())
+        {
+            assert(false);
+            return;
+        }
+
+        if (clioPeers_.size() > 0)
+        {
+            boost::asio::spawn(ioContext_.get(), [this, seq](boost::asio::yield_context yield) {
+                for (auto const& peer : clioPeers_)
+                {
+                    // returns true on success
+                    if (loadCacheFromClioPeer(seq, peer.ip, std::to_string(peer.port), yield))
+                        return;
+                }
+
+                // if we couldn't successfully load from any peers, load from db
+                loadCacheFromDb(seq);
+            });
+            return;
+        }
+        else
+        {
+            loadCacheFromDb(seq);
+        }
+
+        // If loading synchronously, poll cache until full
+        while (cacheLoadStyle_ == LoadStyle::SYNC && not cache_.get().isFull())
+        {
+            log_.debug() << "Cache not full. Cache size = " << cache_.get().size() << ". Sleeping ...";
+            std::this_thread::sleep_for(std::chrono::seconds(10));
+            log_.info() << "Cache is full. Cache size = " << cache_.get().size();
```

Comment:
I think this code just copy from old clio, but the log content just not make sense, right?